File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        result = 0
        min_val, max_val = prices[0], prices[0]
        for price in prices:
            if price > max_val:
                max_val = price
            elif price < min_val:
                if (new_result := max_val - min_val) > result:
                    result = new_result
                min_val = max_val = price
        if (new_result := max_val - min_val) > result:
            return new_result
        else: 
            return result

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int low = prices[0], ans = 0;
        for(int i = 0; i < prices.size(); ++i){
            low = min(low, prices[i]);
            ans = max(ans, prices[i] - low);
        }
        return ans;
    }
};